### DISCUSSION TITLE
DependencyValue keypath does not conform to sendable

### DISCUSSION THREAD

**Discussion by BrentMifsud** [pointfreeco/swift-dependencies#267] [Q&A] [answered]
In the process of swapping out cleanse for swift-dependencies (now that cleanse is archived).

I've been running into this sensibility warning quite a bit while using swift-dependencies.

<img width="1103" alt="image" src="https://github.com/user-attachments/assets/4b58c255-96bb-4660-9d6f-1cfd09ce0ac2">

Has anyone else run into this? Any workarounds?

**Comment by mbrandonw** [comment, marked as answer]
Hi @BrentMifsud, this is just a bug in Swift <6 and will be fixed once Xcode 16 is officially released. For now you can ignore the warning.

**Reply by BrentMifsud** [reply]
> Hi @BrentMifsud, this is just a bug in Swift <6 and will be fixed once Xcode 16 is officially released. For now you can ignore the warning.

Good to know no action required on my end.

Thanks again for putting this library together. So far, it's been dead simple to work with. Love how it's almost identical to the SwiftUI environment.

Random side question:

Any plans to roll out your own `@Entry` macro equivalent? For DependencyValues?

**Reply by mbrandonw** [reply]
> Any plans to roll out your own `@Entry` macro equivalent? For DependencyValues?

There is a PR for part of this (#257) but we're honestly not sure it carries its weight. You can read the discussion [here](https://github.com/pointfreeco/swift-dependencies/discussions/258), and please do chime in if you have some ideas to share.

**Comment by BrentMifsud** [comment]
Seems like this warning still exists in Xcode 16.

**Reply by stephencelis** [reply]
@BrentMifsud You must either be in Swift 6 language mode (which is still not the default and you must opt into it), or you must enable the `InferSendableFromCaptures` upcoming feature from your project's build settings or package's Swift settings.

**Reply by gsabran** [reply]
I've enabled `InferSendableFromCaptures` in my Swift Package and still get the issue while using Swift 5.

**Reply by gsabran** [reply]
Compiling with Swift 6 also gives the warning / error

**Comment by mnddm** [comment]
Hello!
I'm trying to follow the [Meet the Composable Architecture](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/meetcomposablearchitecture/) tutorials and faced with a similar issue. When I added `@Dependency(\.continuousClock) var clock` and tried to use it as a timer, I got the error:

> Capture of 'self' with non-sendable type 'CounterFeature' in a `@Sendable` closure

![error](https://github.com/user-attachments/assets/c0e9ad1d-57b2-4390-b618-e72f18908ba2)

I have Xcode 16 with Swift 6 enabled for my project. I also have a local Swift Package (similar to **isowords** project structure) where I try to place all my code and the Package also has Swift 6.0 enabled. Enabling the `InferSendableFromCaptures` option (in Build Settings or in Package settings) doesn't help, moreover it even raises another error that I'm already using Swift 6.0 with this feature.

I believe conforming the `CounterFeature` to `Sendable` protocol is wrong way. Am I missing something, could you please help? 

Thank you.

**Reply by BrentMifsud** [reply]
> Hello!
> 
> I'm trying to follow the [Meet the Composable Architecture](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/meetcomposablearchitecture/) tutorials and faced with a similar issue. When I added `@Dependency(\.continuousClock) var clock` and tried to use it as a timer, I got the error:
> 
> 
> 
> > Capture of 'self' with non-sendable type 'CounterFeature' in a `@Sendable` closure
> 
> 
> 
> ![error](https://github.com/user-attachments/assets/c0e9ad1d-57b2-4390-b618-e72f18908ba2)
> 
> 
> 
> I have Xcode 16 with Swift 6 enabled for my project. I also have a local Swift Package (similar to **isowords** project structure) where I try to place all my code and the Package also has Swift 6.0 enabled. Enabling the `InferSendableFromCaptures` option (in Build Settings or in Package settings) doesn't help, moreover it even raises another error that I'm already using Swift 6.0 with this feature.
> 
> 
> 
> I believe conforming the `CounterFeature` to `Sendable` protocol is wrong way. Am I missing something, could you please help? 
> 
> 
> 
> Thank you.

I think maybe the swift 6 language mode might have some issues.

I tried turning it on, and despite all the concurrency warnings I still have, I was able to build and compile without any issues.

**Reply by stephencelis** [reply]
Conforming `CounterFeature` to `Sendable` is just fine! It will actually infer it for you for types that aren't exported publicly from the module. So if you remove `public` from this feature, it should build without the explicit conformance, but also explicitly adding the conformance is exactly what you should do when the type is public.

**Comment by BrentMifsud** [comment]
Seems with Xcode 26 this is still a problem. The new default actor isolation feature seems to not play nicely with this.

<img width="860" height="139" alt="image" src="https://github.com/user-attachments/assets/43559289-c053-4a33-854b-e036f3ffe6b5" />

<img width="665" height="80" alt="image" src="https://github.com/user-attachments/assets/bfab4548-fa9c-4990-bdb9-e4b426c17f6a" />

**Reply by BrentMifsud** [reply]
The only fix I've managed to find for this is to mark all my `DependencyKey` and `DependencyValues` extensions `nonisolated`.

The side effect is I've had to give all my isolated dependencies a `nonisolated` initializer.

**Reply by stephencelis** [reply]
This is a different problem. Key paths from actor-isolated scopes are not sendable, and so if you are using default main actor isolation you must explicitly mark things `nonisolated`.

Main actor isolation does seem like a moving target, despite it being the default in Xcode 26, and there are critical fixes that haven't yet made it into the betas. It might be a good idea to weigh into the forums discussions to share your experience with issues you encounter. The goal of default main actor isolation was to simplify the concurrency model, but in practice it seems to complicate things in other ways.

**Reply by samdogg7** [reply]
Any news on this? I am running into this issue myself

**Reply by BrentMifsud** [reply]
> Any news on this? I am running into this issue myself

I was able to solve this by making my DependencyValue extensions static let or computed.

**Reply by mbrandonw** [reply]
@samdogg7 It's always helpful to provide a project that reproduces the problem you are seeing. These errors can have many different underlying sources that are distinct, and now that Swift essentially has completely different dialects via build settings it is even harder to diagnose from a description alone.

**Reply by samdogg7** [reply]
I'll pull together an example and share! But it sounds like `static let` will do the trick, mine were instance properties

**Reply by jaroslavas** [reply]
Maybe you can post a little code snippet of the workaround?

**Reply by BrentMifsud** [reply]
> Maybe you can post a little code snippet of the workaround?

The workaround I mentioned?

```
// thread safe
static let liveValue: MyDependency = MyDependency()

// or

static var liveValue: MyDependency { MyDependency() }

// not thread safe

static var liveValue: MyDependency = MyDependency()

```

**Reply by kesprit** [reply]
> Seems with Xcode 26 this is still a problem. The new default actor isolation feature seems to not play nicely with this.
> 
> <img alt="image" width="860" height="139" src="https://private-user-images.githubusercontent.com/18543934/470627128-43559289-c053-4a33-854b-e036f3ffe6b5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA3OTk0NDYsIm5iZiI6MTc2MDc5OTE0NiwicGF0aCI6Ii8xODU0MzkzNC80NzA2MjcxMjgtNDM1NTkyODktYzA1My00YTMzLTg1NGItZTAzNmYzZmZlNmI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE4VDE0NTIyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxOWFjOTJiZGMyZDlhNjVkYmY4MzIyMWQwZWExOGI3NGVmYjRmZjEwM2Q0YWJmY2NjYzYzOGU5OTA0YWFmZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._jVtFfOWhBi0o5zpDGmVhh9J0I3aV3X1AaTtnj1vOlc"> <img alt="image" width="665" height="80" src="https://private-user-images.githubusercontent.com/18543934/470627245-bfab4548-fa9c-4990-bdb9-e4b426c17f6a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA3OTk0NDYsIm5iZiI6MTc2MDc5OTE0NiwicGF0aCI6Ii8xODU0MzkzNC80NzA2MjcyNDUtYmZhYjQ1NDgtZmE5Yy00OTkwLWJkYjktZTRiNDI2YzE3ZjZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE4VDE0NTIyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI2ZGJhOGU4ZjYxYTE3NGMzM2FiNDg4MTg3ZWRjNTNkM2U5MmU2ZmRlMzFkZmFiZTEyZGU0OWY2MzgwMjM5ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.U13ZK6OOV3ZJ8jtwVcVJkcEWGEZ2rqTggPt_l0NYwQU">

I have the same issue, I just followed the [Composable Architecture Tutorial](https://swiftpackageindex.com/pointfreeco/swift-composable-architecture/main/tutorials/composablearchitecture/01-03-testingyourfeature), I share you my code:

### CounterFeature.swift
```swift
import ComposableArchitecture
import Foundation

@Reducer
struct CounterFeature {
    @ObservableState
    struct State: Equatable {
        var count = 0
        var fact: String?
        var isLoading = false
        var isTimerRunning = false
    }
    
    enum Action {
        case decrementButtonTapped
        case incrementButtonTapped
        case factButtonTapped
        case factResponse(String)
        case toggleTimerButtonTapped
        case timerTick
    }
    
    @Dependency(\.continuousClock) var clock
    @Dependency(\.numberFact) var numberFact
    
    var body: some Reducer<State, Action> {
        Reduce { state, action in
            switch action {
            case .decrementButtonTapped:
                state.count -= 1
                state.fact = nil
                return .none
            case .incrementButtonTapped:
                state.count += 1
                state.fact = nil
                return .none
            case .factButtonTapped:
                state.fact = nil
                state.isLoading = true
                return .run { [count = state.count] send in
                    try await send(.factResponse(self.numberFact.fetch(count)))
                }
            case let .factResponse(fact):
                state.fact = fact
                state.isLoading = false
                return .none
            case .toggleTimerButtonTapped:
                state.isTimerRunning.toggle()
                if state.isTimerRunning {
                    return .run { send in
                        for await _ in await self.clock.timer(interval: .seconds(1)) {
                            await send(.timerTick)
                        }
                    }
                    .cancellable(id: CancelID.timer)
                } else {
                    return .cancel(id: CancelID.timer)
                }
            case .timerTick:
                state.count += 1
                state.fact = nil
                return .none
            }
        }
    }
}

public enum CancelID: String {
    case timer
}
```
### NumberFactClient.swift
```swift
import ComposableArchitecture
import Foundation

struct NumberFactClient {
  var fetch: @Sendable (Int) async throws -> String
}

extension NumberFactClient: DependencyKey {
  static let liveValue = Self(
    fetch: { number in
      let (data, _) = try await URLSession.shared
        .data(from: URL(string: "http://numbersapi.com/\(number)")!)
      return String(decoding: data, as: UTF8.self)
    }
  )
}

extension DependencyValues {
  var numberFact: NumberFactClient {
    get { self[NumberFactClient.self] }
    set { self[NumberFactClient.self] = newValue }
  }
}
```
<img width="1328" height="226" alt="Capture d’écran 2025-10-18 à 17 03 20" src="https://github.com/user-attachments/assets/402781de-77bd-448f-b541-bef732b14c88" />

Can you help me?

**Reply by mbrandonw** [reply]
@kesprit as mentioned above, please provide a full project, not just code snippets. Xcode 26 / Swift 6.2 now comes with entirely different compilation modes and so these errors can largely depend on how your project is set up.

**Reply by kesprit** [reply]
[TCAtuto.zip](https://github.com/user-attachments/files/22985943/TCAtuto.zip)
@mbrandonw here's my project, sorry

**Reply by mbrandonw** [reply]
Thanks for that. The problem is that your project is using default main actor isolation, which is unfortunately the default in new Xcode 26 projects. I personally don't really recommend using that setting as I find it largely gets in your way rather than simplifying things. And you are probably going to have a problem with every 3rd party library with that setting, and even many of Apple's own frameworks don't play nicely with that setting.

But, if you really do want to use that setting, you will find that you often need to mark things as `nonisolated`. For example, the error you are getting right now is because the `DependencyValues.numberFact` property you created is main actor isolated and main actor key paths are never `Sendable`.

So, the fix is to make it non-isolated:

```swift
extension DependencyValues {
  nonisolated var numberFact: NumberFactClient {
    …
  }
}
```

**Reply by kesprit** [reply]
Thanks a lot for the explenation @mbrandonw it's very clear, I'll follow your advice and change the settings.
Both solutions work for me.
For people who will read this, the setting to change is "Default Actor Isolation" -> nonisolated
